File: yaf-gadomain/src/main/java/ar/edu/ungs/yamiko/ga/domain/impl/DynamicLengthGenome.java
package ar.edu.ungs.yamiko.ga.domain.impl;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import ar.edu.ungs.yamiko.ga.domain.Gene;
import ar.edu.ungs.yamiko.ga.domain.Genome;
import ar.edu.ungs.yamiko.ga.domain.Ribosome;

/**
 * Implementación de Genoma de longitud dinámica de tipo T.
 * @author ricardo
 *
 */
public class DynamicLengthGenome<T> implements Genome<T>{

	private Map<String,List<Gene>> structure;
	private int size;
	private Map<Gene,Ribosome<T>> translators;

	public DynamicLengthGenome(String uniqueCrhomosomeName,Gene repetitiveGenes, Ribosome<T> ribosome,int maxLength) 
	{
		structure=new HashMap<String, List<Gene>>();
		List<Gene> genes=new ArrayList<Gene>();
		for (int i=0;i<maxLength;i++)
		{
			Gene x;
			try {
				x = (Gene)repetitiveGenes.clone();
			} catch (CloneNotSupportedException e) {
				x=new BasicGene("Unnamed", 0, 0);
			}
			x.setName(x.getName()+"[" + i + "]");
			genes.add(x);
		}				
		structure.put(uniqueCrhomosomeName,genes);
		size=maxLength;
		translators=new HashMap<Gene,Ribosome<T>>();
		translators.put(repetitiveGenes, ribosome);
	}
	
	public Map<String, List<Gene>> getStructure() {
		return structure;
		}
	
	public int size() {
		return size;
	}
		
	public Map<Gene,Ribosome<T>> getTranslators() {
			return translators;
		}
}

File: yaf-gadomain-scala/src/main/scala/ar/edu/ungs/yamiko/ga/domain/impl/DynamicLengthGenome.scala
package ar.edu.ungs.yamiko.ga.domain.impl

import ar.edu.ungs.yamiko.ga.domain.Genome
import ar.edu.ungs.yamiko.ga.domain.Gene
import ar.edu.ungs.yamiko.ga.domain.Ribosome
import scala.collection.mutable.ListBuffer

@SerialVersionUID(3119L)
class DynamicLengthGenome[T](uniqueCrhomosomeName:String,repetitiveGenes:Gene , ribosome:Ribosome[T],maxLength:Int) extends Genome[T] {

    var structure=Map[String, List[Gene]]()
		var genes=ListBuffer[Gene]();
		for (i<-0 to maxLength-1)
		{
			val x = repetitiveGenes.cloneIt()
			x.setName(x.getName()+"[" + i + "]");
			genes+=x
		}				
		structure+=(uniqueCrhomosomeName -> genes.toList)
		val translators=Map(repetitiveGenes -> ribosome)

	  override def size():Int=maxLength
	  override def getStructure():Map[String,List[Gene]]=structure
	  override def getTranslators():Map[Gene,Ribosome[T]]=translators

}